### ISSUE TITLE
Scaling of parallelizing execute

### ISSUE THREAD

**Issue by giuliojiang** [HPCE/hpce-2017-cw4#21]
Even with very a very large layer (like 40000 in 40000 out which corresponds to a layer.bin over 1GB) the execute:simple only takes about 350ms, so parallelizing the 2 loops in that function won't do much else than adding overhead. In fact if the input data is larger than the network's input size, multiple calls to execute are made each with the right amount of data, and each of these calls would create threads and synchronize and the usual things...

My point is: is this an expected aspect of this particular program? Is it like this to demonstrate that sometimes pipeline parallelization can be far more effective than loop parallelism?

**Comment by m8pple** [comment]
In this case, pipeline parallelism has the bigger advantage, but it depends a lot on the size
of the network and number of layers as to which method is best. You can find a configuration
that supports almost any argument.

So your instinct that you're adding overhead is quite correct, but bear in mind the fairly
awesome size of the numbers we're playing with here: if we assume a (rubbish)
CPU that can to 1GInstr/Sec, then in 350 ms we can get ~350 MInstrs done.

The Intel TBB [guidance on grain size](https://software.intel.com/en-us/node/506060) suggests
that tasks should be 10000-100000 instructions long, so if we take the upper bound
of 100000, then 350 ms gives us time to do about 3500 back-to-back tasks. That is loads
of space in which to soak up some parallelism. In reality the actual network is probably
only taking up ~50-100ms (as you have the data loading), but that is still lots of space.

That said, I wouldn't expect too much of the naive parfor strategy. But if you transform the
problem you can get somewhere. For example, if I take a smaller 4000 -> 4000 layer,
and put through 40000000 bytes (i.e. 10000 samples):
```
dt10@joxer MINGW64 /e/dt10_all/teaching/2017/hpce/hpce-2017-cw4-dt10
$ time < w/random40000000.bin bin/run_network.exe w/layer_4000.bin:simple > w/out.bin
 Loading layer 1 from file w/layer_4000.bin using engine simple

real    0m17.893s
user    0m0.000s
sys     0m0.000s

dt10@joxer MINGW64 /e/dt10_all/teaching/2017/hpce/hpce-2017-cw4-dt10
$ time < w/random40000000.bin bin/run_network.exe w/layer_4000.bin:par_for_clustered > w/out.bin
 Loading layer 1 from file w/layer_4000.bin using engine par_for_clustered

real    0m4.785s
user    0m0.000s
sys     0m0.015s
```

Though for 40000 I actually see slower performance... All kinds of effects such as cache
come into play as this is a very memory intensive benchmark. This application exhibits
(by design) fairly complex behaviour, and you would have to tune your parallelism
strategy to both the network and the architecture if you want the best performance.
Real-world neural net libraries like TensorFlow spend quite a lot of time trying to
optimise the execution graph, and take a lot of care about how to map different
parts of the graphs to parallel components.

**Comment by giuliojiang** [comment]
Yes, I forgot to mention that the implementation that shows really high overhead is `par_for_naive`, while `par_for_clustered` performs exceptionally well with any layer size above 512-512.

Is there any particular reason why the clustered version seems to be so much more efficient than the naive version? The naive parfor could maybe have some cache trashing issues as both src and destination positions are unpredictable when iterating through the synapses, but then the clustered parfor would only solve this for the output layer. Maybe the fact that every task is only writing to a single output cell makes it good enough for the cache to be used effectively?

**Comment by m8pple** [comment]
As you say, the naive version is trashing the cache, and those writes to the destination
layer are particularly painful - writes to RAM are more expensive than reads (in a multicore
context), a read-modify-write is even more expensive, and a read-modify-write to a location
that other CPUs might modify is the most expensive of all.

If you look at the naive version, you estimate the per-iteration cost of the first loop as
- one read-only predictable (linear pattern) accesses to `m_synapses` (assuming three reads are merged)
- one read-only unpredictable access to `pIn`
- one read-modify-write unpredictable to `pOut`

That read-modify-write is the performance killer here, as you are doing `nSynapses` of them,
and every task is trying to read-modify-write random points in the same array. So you'll have different
CPUs fighting for the same cache lines, which makes every access much more expensive.

If you then count the accesses in the clustered version, you'll (hopefully) see that there are actually
no read-modify-writes at all. Data may be read more than once (e.g. from `pIn`), but each
output item is written exactly once. You can also set things up so that all the writes are
adjacent to each other, so each cache line is "owned" by one CPU as it fills it up. So ideally
most of the data will be either read-only shared (`pIn`), read-only private (the synapse data),
or write-only private (`pOut`).

**Comment by giuliojiang** [comment]
Thanks for the answer. It's very useful!
